### DISCUSSION TITLE
cohortgenerator- extract all vaccine dates as a string of dates

### DISCUSSION THREAD

**Discussion by p-stehlik** [opensafely/documentation#1323] [Q&A]
Hi OpenSafely peeps

I'm wanting to extract all the dates of a patients previous COVID vaccinations and positive test results and classify their immunity level according to these.

Is there a way to do this with the cohortgenerator?

Thank you

**Comment by alexwalkerepi** [comment]
Hello! OpenSAFELY was designed initially to return data in the form of one row per patient. This was because it met the early, urgent needs of using the data during the pandemic, and also minimised the amount of data returned, which is another layer of privacy protection. However since then we've started using it for more complex studies where event level data like that you describe would be useful. There are plans to implement this functionality at some point in the future, but I'm not sure of the timescale.

In the mean time, there's an imperfect but okay workaround we've developed to return multiple events, where you end up with wide data rather than long. It involves:
1. selecting the first event of interest
2. using the date of that first event to find the next event (using `on_or_after` in cohortextractor)
3. using the date of that second event to find the next one
4. and so on...

This is fine if you're only expecting a small number of events, but if you need more then a few tens it will take a long time to run the query and return an enormous output file. To write it less laboriously you can do it in a loop (example [here](https://github.com/opensafely/long-covid-sick-notes/blob/b880b26bd9b029e4e9503ab70f4619ade2b31f15/analysis/common_variables.py#L42-L105)).

P.S. [cohortextractor ](https://docs.opensafely.org/study-def/) is due to be deprecated within the next few months, and we expect all new studies will use it's new, more flexible and better designed cousin [ehrQL](https://docs.opensafely.org/ehrql/), which will hopefully eventually have a better way of working with event level data.

**Comment by p-stehlik** [comment]
Hi Alex

I’d thought about that work around but decided it would be a burdensome.

When you point to the loop do you mean:

**{
            f"sick_note_{n}_duration_months": patients.with_these_clinical_events(
                codelist(["Y08c2"], system="ctv3"),
                between=[f"sick_note_{n}_date", f"sick_note_{n}_date"],
                returning="numeric_value",
                return_expectations={
                    "incidence": 0.9,
                    "float": {"distribution": "normal", "mean": 21, "stddev": 5},
                },
            )
            for n in range(1, 6)
        },
        **covariates,

I assume this will create variables that get given a similar name from 1-6?
If so this would be very useful to as a temporary fix – thank you.

I’ll explore the ehrQL<https://protect-au.mimecast.com/s/0Vr7COMxVEHAzRBlOik_SLc?domain=docs.opensafely.org/> tool a bit more – are the some GitHub examples with its use yet?

Thanks in advance.

Kind regards,
Paulie
